### PR TITLE
Adding glob matching pattern

### PR DESCRIPTION
**Pull request by Ofir-Lapid**

### CODE REVIEW

File: kubebeat/beater/file_path_x.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+				})
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+		matches = hits
+	}
+
+	// fix up return value for nil input
```
Comment (Ofir-Lapid) [inline kubebeat/beater/file_path_x.go:55]: Maybe we can allocate an empty array instead

File: kubebeat/beater/file_system_fetcher.go [hunk 2]
```diff
@@ -1,19 +1,18 @@
...
-
-	"github.com/elastic/beats/v7/libbeat/logp"
 )
 
 // FileSystemFetcher implement the Fetcher interface
 // The FileSystemFetcher meant to fetch file/directories from the file system and ship it
 // to the Kubebeat
 type FileSystemFetcher struct {
-	filesPaths []string // Files and directories paths for the fetcher to extract info from
+	inputFiles []string // Files and directories paths for the fetcher to extract info from
```
Comment (yashtewari) [inline kubebeat/beater/file_system_fetcher.go:15]: Maybe something like `filePatterns` or `inputFilePatterns`?

**Review by yashtewari — APPROVED** [review]
Looks great!

**Review by amirbenun — COMMENTED** [review]
In the current implementation, I would expect more test coverage on `glob_mathicng.go`.
Tests on `file_system_fetcher.go` should cover new functionality introduced there which is edge cases, handling with several patterns, the structure of each result.

File: kubebeat/beater/glob_matching.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
+package beater
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+
+// Based on https://github.com/yargevad/filepathx/blob/master/filepathx.go
```
[the remark this answers is not in the document]
Comment (Ofir-Lapid) [reply kubebeat/beater/glob_matching.go:10]: I thought about using the package directly but the repo doesn't has that much of starts, so I didn't feel that we should rely on that repo (In case it will ever change). Amir also agreed with that.

WDT ?

[the remark this answers is not in the document]
Comment (yashtewari) [reply kubebeat/beater/glob_matching.go:10]: Yeah, sound reasoning. In that case, please bring the tests over, they're free!

File: kubebeat/beater/file_system_fetcher.go (hunk 2 above)
Comment (yashtewari) [reply kubebeat/beater/file_system_fetcher.go:15]: It's safe to say that a directory is just a special kind of file. But if you can also go with something like `pathPatterns`. My point is that "pattern" should be explicit here.

File: kubebeat/beater/file_system_fetcher.go [hunk 4]
```diff
@@ -33,30 +32,37 @@ type FileSystemResourceData struct {
 
 func NewFileFetcher(filesPaths []string) Fetcher {
 	return &FileSystemFetcher{
-		filesPaths: filesPaths,
+		inputFilePatterns: filesPaths,
```
Comment (eyalkraft) [inline kubebeat/beater/file_system_fetcher.go:35]: maybe also rename the argument (`filePatterns`?)

**Review by eyalkraft — APPROVED** [review]
LGTM

File: kubebeat/beater/glob_matching_test.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
+package beater
```
Comment (amirbenun) [inline kubebeat/beater/glob_matching_test.go:1]: liked the tests, please add a tests of a pattern that doesn't exist though.

File: kubebeat/beater/glob_matching_test.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+	if err != nil {
+		assert.Fail(t, "Glob could not fetch results", err)
+	}
+
+	assert.Equal(t, len(matchedFiles), 6)
+
+	//When using glob matching recursively the first outer folder is being sent without a '/'
+	assert.Equal(t, matchedFiles[0], outerDir + "/")
+	assert.Equal(t, matchedFiles[1], filepath.Join(outerDir, outerFiles[0]))
```
Comment (amirbenun) [inline kubebeat/beater/glob_matching_test.go:90]: Please note `func Equal(t TestingT, expected, actual interface{}, msgAndArgs ...interface{}) bool`.
You should switch the order of your parameters being pass (globally on the pr).
Also I think the expected argument should be a raw string parameter and not calculated value like `filepath.Join` result.
That will make the test much more readable.

File: kubebeat/beater/file_system_fetcher_test.go [hunk 7]
```diff
@@ -9,28 +9,147 @@ import (
...
+
+	innerInnerDirectoryName := "test-inner-inner-dir"
+	innerInnerFiles := []string{"innerInnerFolderFile.txt"}
+	innerInnerDir := createDirectoriesWithFiles(t, innerDir, innerInnerDirectoryName, innerInnerFiles)
+
+	path := []string{outerDir + "/**"}
+	fileFetcher := NewFileFetcher(path)
+	results, err := fileFetcher.Fetch()
+
+	if err != nil {
```
Comment (amirbenun) [inline kubebeat/beater/file_system_fetcher_test.go:125]: There should be an `assert.NotError` or `assert.Nil` instead of that if statement

**Review by amirbenun — COMMENTED** [review]
A few comments regarding the tests

File: kubebeat/beater/glob_matching_test.go (hunk 5 above)
Comment (Ofir-Lapid) [reply kubebeat/beater/glob_matching_test.go:1]: Done

File: kubebeat/beater/glob_matching_test.go (hunk 6 above)
Comment (Ofir-Lapid) [reply kubebeat/beater/glob_matching_test.go:90]: I cannot use a raw string since the outdir path is being autogenerated.

As for the first note - done

**Review by amirbenun — APPROVED** [review]
